src/app/Dashboard/Charts/MBeanMetricsChartController.tsx

```diff
@@ -0,0 +1,225 @@
+        }
+        if (prev && !curr) {
+          // last subscriber left
+          this._stop();
+        }
+      });
+  }
+
+  attach(category: string, fields: string[]): Observable<MBeanMetrics> {
+    this._refCount$.next(this._refCount$.value + 1);
+    if (!this._metrics.has(category)) {
+      this._metrics.set(category, []);
+    }
+    this._metrics.get(category)?.push(...fields);
+    return this._state$.asObservable().pipe(
+      distinctUntilChanged(),
+      finalize(() => {
+        this._refCount$.next(this._refCount$.value - 1);
+        const original = this._metrics.get(category) || [];
+        const updated = [...original];
+        fields.forEach((field) => {
+          const idx = updated.findIndex((e) => e === field);
+          if (idx >= 0) {
+            updated.splice(idx, 1);
+          }
+        });
+        this._metrics.get(category)?.push(...updated);
+      })
+    );
+  }
+
+  requestRefresh(): void {
+    this._updates$.next();
+  }
+
+  loading(): Observable<boolean> {
+    return this._loading$.asObservable().pipe(distinctUntilChanged());
+  }
+
+  _tearDown() {
+    this._state$.next({});
+    this._lazy.unsubscribe();
+    this._stop();
+  }
+
+  private _stop(): void {
+    if (this._attach) {
+      this._attach.unsubscribe();
+      this._attach = undefined;
+    }
+  }
+
+  private _start(): void {
+    this._stop();
+    this._attach = merge(
+      this._updates$.pipe(
+        throttleTime(this._settings.chartControllerConfig().minRefresh * 1000),
+        concatMap((_) => this._target.target().pipe(first()))
+      ),
+      this._target.target()
+    )
+      .pipe(
+        tap((_) => this._loading$.next(true)),
+        concatMap((t) => this._queryMetrics(t)),
+        tap((_) => this._loading$.next(false))
+      )
+      .subscribe((v) => this._state$.next(v));
+  }
+
+  private _queryMetrics(target: Target): Observable<MBeanMetrics> {
+    const q: string[] = [];
+    const m = new Map<string, Set<string>>();
+    this._metrics.forEach((fields, category) => {
+      const s = m.get(category) || new Set<string>();
+      fields.forEach((f) => s.add(f));
+      m.set(category, s);
+    });
+    m.forEach((s, k) => {
+      let l = `${k} {`;
+      s.forEach((f) => (l += `\t${f}\n`));
+      l += '}';
+      q.push(l);
+    });
+    return this._api
+      .graphql<MBeanMetricsResponse>(
+        `
+          query MBeanMXMetricsForTarget($connectUrl: String) {
+            targetNodes(filter: { name: $connectUrl }) {
+              mbeanMetrics {
+                ${q.join('\n')}
+              }
+            }
+          }`,
+        { connectUrl: target.connectUrl }
+      )
+      .pipe(
+        map((resp) => {
+          const nodes = resp.data.targetNodes;
+          if (!nodes || nodes.length === 0) {
+            return {};
+          }
+          return nodes[0]?.mbeanMetrics;
+        }),
+        catchError((_) => of({}))
+      );
+  }
+}
+
+export interface MemoryMetric {
```

Comment:
Maybe should be called `MemoryUsage` or `MemoryUsageMetric` to match the java object?